src/tensors/cpu/ruy_adapter.h

```diff
@@ -0,0 +1,468 @@
+/*
+ * This file follows intgemm and is a means of retrofitting ruy into the intgemm based wiring in
+ * `intgemm_interface.h`. ruy is an inference backend used in tensorflow and android deployment and
+ * has an optimized ARM backend for the multiply operations required. Optimized code for quantize,
+ * unquantize, transpose are added separately to connect the multiply library to marian.
+ */
+
+#pragma once
+#include <algorithm>
+#include <cmath>
+#include <cstdint>
+#include "ruy/platform.h"
+#include "ruy/system_aligned_alloc.h"
+
+#if RUY_PLATFORM_NEON
+#include <arm_neon.h>
+#endif
+
+namespace marian {
+namespace cpu {
+namespace integer {
+
+using Index = unsigned int;
+
+// The following partitions a pure C++ slow implementation and a faster SIMD implementation using
+// NEON intrinsics on ARM hardware. Ruy already has such a routing, but we add some preprocessing
+// and postprocessing functions (quantize, transpose, unquantize) that are outside ruy's offerings
+// and required in the fast matrix-multiplication workflow for machine-translation, that exists in
+// marian.
+
+enum class Path {
+  kStandardCpp = 0,  // Pure C++
+  kNeon = 1          // NEON Intrinsics (ARM)
+};
+
+#if RUY_PLATFORM_NEON
+constexpr Path kHighestPath = Path::kNeon;
+#else
+constexpr Path kHighestPath = Path::kStandardCpp;
+#endif
+
+template <enum Path>
+struct Preprocess;
+
+/*
+ * Naive implementation using standard C++ functions. Not optimized using SIMD operations.
+ */
+template <>
+struct Preprocess<Path::kStandardCpp> {
```

Comment:
First impression would be just #ifdef the implementation of preprocess since it's coming from a macro anyway.  If you want to have the reference implementation, maybe call this PreprocessReference, then later
```
#ifdef RUY_PLATFORM_NEON
class Preprocess {
 // neon implementation
};
#else
typedef PreprocessReference Preprocess;
#endif
```

Reply:
Reference is removed altogether. There exists only neon implementation now. `Preprocess<Path>` namespace is replaced by free functions that are currently inline.